### DISCUSSION TITLE
Allow flexibility with flag placement

### DISCUSSION THREAD

**Discussion by philomory** [helmfile/helmfile#185] [Ideas]
There is an awkward quirk to the way that `helmfile` accepts flags that occasionally still catches me off-guard even after using it extensively for a while, which can be most quickly shown with an example: `helmfile apply --debug` is invalid. So is `helmfile diff -e test`. More generally, any flag associated with a particular sub-command _must_ appear _after_ that subcommand on the command line, whereas any flag that is global to helmfile as a whole such as `--debug` or `-e` _must_ appear _before_ any subcommands.

This used to trip me up constantly; these days it doesn't get me _as_ often, but it's still a bit annoying if you run a complex command with a lot of flags, and it fails, that you can't just press the up arrow, type ` --debug` and hit enter, because you'll just get `flag provided by not defined: -debug`; instead you have to press up to retrieve the previous command from history, then carefully navigate your cursor to the "correct" part of the command (before the subcommand) before adding your `--debug` flag.

It also comes up when you'd like to run the same command in multiple environments, or with multiple selectors; it'd be nice to run e.g. `helmfile diff -e test -l name=foo`, and then be able to just press up, backspace a few times, type a different selector, and press enter, rather than having to faff about with the arrow keys.

**Comment by philomory** [comment]
I'll note that other tools in the ecosystem work as I'm describing; for example, `kubectl -n foo get pods` is just as valid as `kubectl get pods -n foo`, and `helm --kube-context foo history bar`, `helm history --kube-context foo bar` and `helm history bar --kube-context foo` are all equivalent.

**Comment by yxxhero** [comment]
@philomory Thanks for your feedback. I will fix this issue.

**Comment by yxxhero** [comment]
@philomory done
